allennlp/data/token_indexers/wordpiece_indexer.py

```diff
@@ -137,39 +179,59 @@ def tokens_to_indices(self,
-                    offset += len(token_wordpiece_ids)
-                # For final offsets, the current value of ``offset`` is the end of
-                # the previous wordpiece, so increment it and then add it to ``offsets``.
-                else:
-                    offset += len(token_wordpiece_ids)
-                    offsets.append(offset)
-                # And add the token_wordpiece_ids to the output list.
-                wordpiece_ids.extend(token_wordpiece_ids)
-            else:
-                # TODO(joelgrus): figure out a better way to handle this
-                logger.warning(f"Too many wordpieces, truncating: {[token.text for token in tokens]}")
+        offset = len(self._start_piece_ids) if self.use_starting_offsets else len(self._start_piece_ids) - 1
+
+        for token in token_wordpiece_ids:
+            # Truncate the sequence if specified, which depends on where the offsets are
+            next_offset = 1 if self.use_starting_offsets else 0
+            if self._truncate_long_sequences and offset >= window_length + next_offset:
                 break
 
-        # By construction, we still have enough room to add the end_token ids.
-        wordpiece_ids.extend(self._end_piece_ids)
-        # Constructing `token_type_ids` by `self._separator`
-        token_type_ids = _get_token_type_ids(wordpiece_ids,
-                                             self._separator_ids)
+            # For initial offsets, the current value of ``offset`` is the start of
+            # the current wordpiece, so add it to ``offsets`` and then increment it.
+            if self.use_starting_offsets:
+                offsets.append(offset)
+                offset += len(token)
+            # For final offsets, the current value of ``offset`` is the end of
+            # the previous wordpiece, so increment it and then add it to ``offsets``.
+            else:
+                offset += len(token)
+                offsets.append(offset)
+
+        if len(flat_wordpiece_ids) <= window_length:
+            # If all the wordpieces fit, then we don't need to do anything special
+            wordpiece_windows = [self._add_start_and_end(flat_wordpiece_ids)]
+            token_type_ids = self._extend(flat_token_type_ids)
+        elif self._truncate_long_sequences:
+            logger.warning("Too many wordpieces, truncating sequence. If you would like a sliding window, set"
+                           "`truncate_long_sequences` to False %s", str([token.text for token in tokens]))
+            wordpiece_windows = [self._add_start_and_end(flat_wordpiece_ids[:window_length])]
+            token_type_ids = self._extend(flat_token_type_ids)
+        else:
+            # Create a sliding window of wordpieces of length `max_pieces` that advances by `stride` steps and
+            # add start/end wordpieces to each window
+            # TODO: this currently does not respect word boundaries, so words may be cut in half between windows
```

Comment:
i'm not sure why this is a problem...am i missing something?

Reply:
Just a small technical consideration. I guess it's less important now since I changed it to maximal context embeddings which use the center embeddings.